cmd/net.go

```diff
@@ -86,20 +86,18 @@ func getHostIP4(host string) (ipList set.StringSet, err error) {
 		startTime := time.Now()
 		// wait for hosts to resolve in exponentialbackoff manner
 		for _ = range newRetryTimerSimple(doneCh) {
-			// Retry infinitely on Kubernetes and Docker swarm.
-			// This is needed as the remote hosts are sometime
-			// not available immediately.
 			if ips, err = net.LookupIP(host); err == nil {
 				break
 			}
 			// time elapsed
 			timeElapsed := time.Since(startTime)
 			// log error only if more than 1s elapsed
-			if timeElapsed > time.Second {
+			if timeElapsed > defaultRetryCap {
 				// log the message to console about the host not being
 				// resolveable.
 				errorIf(err, "Unable to resolve host %s (%s)", host,
 					humanize.RelTime(startTime, startTime.Add(timeElapsed), "elapsed", ""))
+				break
```

Comment:
Thanks for the PR.   Unfortunately we cannot actually do this @ws141 - there is a reason why getHostIP4 is infinitely retrying.  We should fix it differently rather than removing the requirement of this function.

Reply:
You can simply do this. 

```diff
diff --git a/cmd/endpoint_test.go b/cmd/endpoint_test.go
index e93dd10..547e421 100644
--- a/cmd/endpoint_test.go
+++ b/cmd/endpoint_test.go
@@ -38,42 +38,46 @@ func TestNewEndpoint(t *testing.T) {
 		expectedEndpoint Endpoint
 		expectedType     EndpointType
 		expectedErr      error
+		skipDocker       bool
 	}{
-		{"foo", Endpoint{URL: &url.URL{Path: "foo"}, IsLocal: true}, PathEndpointType, nil},
-		{"/foo", Endpoint{URL: &url.URL{Path: "/foo"}, IsLocal: true}, PathEndpointType, nil},
-		{`\foo`, Endpoint{URL: &url.URL{Path: `\foo`}, IsLocal: true}, PathEndpointType, nil},
-		{"C", Endpoint{URL: &url.URL{Path: `C`}, IsLocal: true}, PathEndpointType, nil},
-		{"C:", Endpoint{URL: &url.URL{Path: `C:`}, IsLocal: true}, PathEndpointType, nil},
-		{"C:/", Endpoint{URL: &url.URL{Path: "C:"}, IsLocal: true}, PathEndpointType, nil},
-		{`C:\`, Endpoint{URL: &url.URL{Path: `C:\`}, IsLocal: true}, PathEndpointType, nil},
-		{`C:\foo`, Endpoint{URL: &url.URL{Path: `C:\foo`}, IsLocal: true}, PathEndpointType, nil},
-		{"C:/foo", Endpoint{URL: &url.URL{Path: "C:/foo"}, IsLocal: true}, PathEndpointType, nil},
-		{`C:\\foo`, Endpoint{URL: &url.URL{Path: `C:\\foo`}, IsLocal: true}, PathEndpointType, nil},
-		{"http:path", Endpoint{URL: &url.URL{Path: "http:path"}, IsLocal: true}, PathEndpointType, nil},
-		{"http:/path", Endpoint{URL: &url.URL{Path: "http:/path"}, IsLocal: true}, PathEndpointType, nil},
-		{"http:///path", Endpoint{URL: &url.URL{Path: "http:/path"}, IsLocal: true}, PathEndpointType, nil},
-		{"http://localhost/path", Endpoint{URL: u1, IsLocal: true}, URLEndpointType, nil},
-		{"http://localhost/path//", Endpoint{URL: u1, IsLocal: true}, URLEndpointType, nil},
-		{"https://example.org/path", Endpoint{URL: u2}, URLEndpointType, nil},
-		{"http://127.0.0.1:8080/path", Endpoint{URL: u3, IsLocal: true}, URLEndpointType, nil},
-		{"http://192.168.253.200/path", Endpoint{URL: u4}, URLEndpointType, nil},
-		{"", Endpoint{}, -1, fmt.Errorf("empty or root endpoint is not supported")},
-		{"/", Endpoint{}, -1, fmt.Errorf("empty or root endpoint is not supported")},
-		{`\`, Endpoint{}, -1, fmt.Errorf("empty or root endpoint is not supported")},
-		{"c://foo", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format")},
-		{"ftp://foo", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format")},
-		{"http://server/path?location", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format")},
-		{"http://:/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: invalid port number")},
-		{"http://:8080/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: empty host name")},
-		{"http://server:/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: invalid port number")},
-		{"https://93.184.216.34:808080/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: port number must be between 1 to 65535")},
-		{"http://server:8080//", Endpoint{}, -1, fmt.Errorf("empty or root path is not supported in URL endpoint")},
-		{"http://server:8080/", Endpoint{}, -1, fmt.Errorf("empty or root path is not supported in URL endpoint")},
-		{"http://server/path", Endpoint{}, -1, fmt.Errorf("lookup server" + errMsg)},
-		{"192.168.1.210:9000", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: missing scheme http or https")},
+		{"foo", Endpoint{URL: &url.URL{Path: "foo"}, IsLocal: true}, PathEndpointType, nil, false},
+		{"/foo", Endpoint{URL: &url.URL{Path: "/foo"}, IsLocal: true}, PathEndpointType, nil, false},
+		{`\foo`, Endpoint{URL: &url.URL{Path: `\foo`}, IsLocal: true}, PathEndpointType, nil, false},
+		{"C", Endpoint{URL: &url.URL{Path: `C`}, IsLocal: true}, PathEndpointType, nil, false},
+		{"C:", Endpoint{URL: &url.URL{Path: `C:`}, IsLocal: true}, PathEndpointType, nil, false},
+		{"C:/", Endpoint{URL: &url.URL{Path: "C:"}, IsLocal: true}, PathEndpointType, nil, false},
+		{`C:\`, Endpoint{URL: &url.URL{Path: `C:\`}, IsLocal: true}, PathEndpointType, nil, false},
+		{`C:\foo`, Endpoint{URL: &url.URL{Path: `C:\foo`}, IsLocal: true}, PathEndpointType, nil, false},
+		{"C:/foo", Endpoint{URL: &url.URL{Path: "C:/foo"}, IsLocal: true}, PathEndpointType, nil, false},
+		{`C:\\foo`, Endpoint{URL: &url.URL{Path: `C:\\foo`}, IsLocal: true}, PathEndpointType, nil, false},
+		{"http:path", Endpoint{URL: &url.URL{Path: "http:path"}, IsLocal: true}, PathEndpointType, nil, false},
+		{"http:/path", Endpoint{URL: &url.URL{Path: "http:/path"}, IsLocal: true}, PathEndpointType, nil, false},
+		{"http:///path", Endpoint{URL: &url.URL{Path: "http:/path"}, IsLocal: true}, PathEndpointType, nil, false},
+		{"http://localhost/path", Endpoint{URL: u1, IsLocal: true}, URLEndpointType, nil, false},
+		{"http://localhost/path//", Endpoint{URL: u1, IsLocal: true}, URLEndpointType, nil, false},
+		{"https://example.org/path", Endpoint{URL: u2}, URLEndpointType, nil, false},
+		{"http://127.0.0.1:8080/path", Endpoint{URL: u3, IsLocal: true}, URLEndpointType, nil, false},
+		{"http://192.168.253.200/path", Endpoint{URL: u4}, URLEndpointType, nil, false},
+		{"", Endpoint{}, -1, fmt.Errorf("empty or root endpoint is not supported"), false},
+		{"/", Endpoint{}, -1, fmt.Errorf("empty or root endpoint is not supported"), false},
+		{`\`, Endpoint{}, -1, fmt.Errorf("empty or root endpoint is not supported"), false},
+		{"c://foo", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format"), false},
+		{"ftp://foo", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format"), false},
+		{"http://server/path?location", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format"), false},
+		{"http://:/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: invalid port number"), false},
+		{"http://:8080/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: empty host name"), false},
+		{"http://server:/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: invalid port number"), false},
+		{"https://93.184.216.34:808080/path", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: port number must be between 1 to 65535"), false},
+		{"http://server:8080//", Endpoint{}, -1, fmt.Errorf("empty or root path is not supported in URL endpoint"), false},
+		{"http://server:8080/", Endpoint{}, -1, fmt.Errorf("empty or root path is not supported in URL endpoint"), false},
+		{"http://server/path", Endpoint{}, -1, fmt.Errorf("lookup server" + errMsg), true},
+		{"192.168.1.210:9000", Endpoint{}, -1, fmt.Errorf("invalid URL endpoint format: missing scheme http or https"), false},
 	}
 
 	for _, testCase := range testCases {
+		if testCase.skipDocker && (IsDocker() || IsKubernetes()) {
+			continue
+		}
 		endpoint, err := NewEndpoint(testCase.arg)
 		if testCase.expectedErr == nil {
 			if err != nil {
diff --git a/cmd/net_test.go b/cmd/net_test.go
index 4dde1d9..d861f91 100644
--- a/cmd/net_test.go
+++ b/cmd/net_test.go
@@ -128,18 +128,22 @@ func TestMustGetLocalIP4(t *testing.T) {
 }
 
 func TestGetHostIP(t *testing.T) {
-	_, err := getHostIP4("myserver")
+	errMsg := ": no such host"
 	testCases := []struct {
 		host           string
 		expectedIPList set.StringSet
 		expectedErr    error
+		skipDocker     bool
 	}{
-		{"localhost", set.CreateStringSet("127.0.0.1"), nil},
-		{"example.org", set.CreateStringSet("93.184.216.34"), nil},
-		{"myserver", nil, err},
+		{"localhost", set.CreateStringSet("127.0.0.1"), nil, false},
+		{"example.org", set.CreateStringSet("93.184.216.34"), nil, false},
+		{"myserver", nil, fmt.Errorf("lookup server" + errMsg), true},
 	}
 
 	for _, testCase := range testCases {
+		if testCase.skipDocker && (IsDocker() || IsKubernetes()){
+			continue
+		}
 		ipList, err := getHostIP4(testCase.host)
 		if testCase.expectedErr == nil {
 			if err != nil {
```

This should work @ws141 .. // cc @NitishT

Reply:
:+1:

Reply:
@harshavardhana Sure, that approach works. Will you address the issue or do you want another PR with this?